firebase-perf/src/main/java/com/google/firebase/perf/application/FrameMetricsRecorder.java

```diff
@@ -0,0 +1,182 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.perf.application;
+
+import android.app.Activity;
+import android.util.SparseIntArray;
+import androidx.core.app.FrameMetricsAggregator;
+import com.google.android.gms.common.util.VisibleForTesting;
+import com.google.firebase.perf.logging.AndroidLogger;
+import com.google.firebase.perf.metrics.FrameMetricsCalculator.PerfFrameMetrics;
+import com.google.firebase.perf.util.Constants;
+import com.google.firebase.perf.util.Optional;
+import java.util.WeakHashMap;
+
+/**
+ * Provides FrameMetrics data from an Activity's Window. Encapsulates FrameMetricsAggregator.
+ *
+ * <p>Note: each recorder holds a reference to an Activity, so it is very important to dereference
+ * each recorder before the associated Activity is destroyed
+ */
+public class FrameMetricsRecorder {
+  private static final AndroidLogger logger = AndroidLogger.getInstance();
+  private final Activity activity;
+  private final FrameMetricsAggregator fma;
+  private final WeakHashMap<Object, PerfFrameMetrics> subTraceMap;
+  private boolean isRecording = false;
+
+  /**
+   * Creates a recorder for a specific activity
+   *
+   * @param activity the activity that the recorder is collecting data from
+   */
+  public FrameMetricsRecorder(Activity activity) {
+    this(activity, new FrameMetricsAggregator(), new WeakHashMap<>());
+  }
+
+  @VisibleForTesting
+  FrameMetricsRecorder(
+      Activity activity,
+      FrameMetricsAggregator frameMetricsAggregator,
+      WeakHashMap<Object, PerfFrameMetrics> subTraceMap) {
+    this.activity = activity;
+    this.fma = frameMetricsAggregator;
+    this.subTraceMap = subTraceMap;
+  }
+
+  /** Starts recording FrameMetrics for the activity window */
+  public void start() {
+    if (isRecording) {
+      logger.error(
+          "FrameMetricsAggregator is already recording %s", activity.getClass().getSimpleName());
+      return;
+    }
+    fma.add(activity);
+    isRecording = true;
+  }
+
+  /**
+   * Stops recording FrameMetrics for the activity window
+   *
+   * @return FrameMetrics accumulated during the current recording
+   */
+  public Optional<PerfFrameMetrics> stop() {
+    if (!isRecording) {
+      logger.error("Cannot stop because no recording was started");
+      return Optional.absent();
+    }
+    Optional<PerfFrameMetrics> data = this.snapshot();
+    try {
+      fma.remove(activity);
+    } catch (IllegalArgumentException err) {
+      logger.debug("View not hardware accelerated. Unable to collect FrameMetrics. %s", err);
```

Comment:
Is this the only reason we can have an IAE?

Reply:
Yes. Android will try to register a listener on the top level view of the Activity Window to listen for frame times. If the view isn't hardware accelerated, the listener will not be registered but no error will occur. Only when trying to remove it, it will throw IAE when that listener can't be found.